src/tests/zfs/zfs_ha_controller/zfs_ha_controller_test.go

```diff
@@ -0,0 +1,258 @@
+package zfs_ha_controller
+
+import (
+	"testing"
+	"time"
+
+	"github.com/openebs/openebs-e2e/common/e2e_config"
+	"github.com/openebs/openebs-e2e/common/e2e_ginkgo"
+	"github.com/openebs/openebs-e2e/common/k8sinstall"
+	"github.com/openebs/openebs-e2e/common/zfs"
+
+	"github.com/openebs/openebs-e2e/common"
+	"github.com/openebs/openebs-e2e/common/k8stest"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	coreV1 "k8s.io/api/core/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+/*
+	Background:
+		Given that the product is installed on a kubernetes cluster
+	Scenario:  Zfs controller in high availability mode
+		Given Zfs controller deployed in non HA mode
+		When Zfs controller is scaled down to zero replica
+		And Zfs volume(pvc) is deployed
+		Then Zfs volume(pvc) should remain in pending state
+		When Zfs controller is scaled to two replica via helm upgrade
+		Then Zfs volume(pvc) should transition to bound state
+		And Deploy fio application to use zfs volume
+		And Taint all worker node with NoSchedule taint
+		When Delete one of the zfs controller pod
+		Then Zfs node lease should point to running zfs controller pod
+*/
+
+var nodeConfig zfs.ZfsNodesDevicePoolConfig
+var defLeaseSwitchTime = 120 // in seconds
+var nodesWithoutTaint []string
+var zfsControllerOrgReplica int32
+var app k8stest.FioApplication
+var volumeProvisionErrorMsg = "waiting for a volume to be created, either by external provisioner"
+
+func controllerHaTest(decor string, engine common.OpenEbsEngine, volType common.VolumeType, fstype common.FileSystemType, volBindModeWait bool) {
+	var err error
+	e2e_config := e2e_config.GetConfig().Product
+
+	app = k8stest.FioApplication{
+		Decor:                          decor,
+		VolSizeMb:                      1024,
+		OpenEbsEngine:                  engine,
+		VolType:                        volType,
+		FsType:                         fstype,
+		Loops:                          5,
+		VolWaitForFirstConsumer:        volBindModeWait,
+		SkipPvcVerificationAfterCreate: true,
+	}
+
+	// setup sc parameters
+	app.Zfs = k8stest.ZfsOptions{
+		PoolName:      nodeConfig.PoolName,
+		ThinProvision: common.No,
+		RecordSize:    "128k",
+		Compression:   common.Off,
+		DedUp:         common.Off,
+	}
+
+	// list nodes without NoSchedule taint
+	nodesWithoutTaint, err = k8stest.ListNodesWithoutNoScheduleTaint()
+	Expect(err).ToNot(HaveOccurred(), "failed to list nodes without NoSchedule taint")
+	logf.Log.Info("nodes without NoSchedule taint", "nodes", nodesWithoutTaint)
+
+	// get the no of replicas in zfs-controller deployment
+	// Scale down the zfs-controller deployment
+	// Check that zfs-controller pods has been terminated successfully
+	logf.Log.Info("Scale down and get the no of replicas in zfs-controller deployment")
+	zfsControllerName := e2e_config.ZfsEngineControllerDeploymentName
+	zfsControllerOrgReplica, err = k8sinstall.ScaleZfsControllerViaHelm(0)
```

Comment:
Please can you add  comments describing behaviour here in more detail. 
Logically, scaling controllers to 0 would imply that there are no controllers present on successful scaling operation. 
However this does not appear to be the case.

Reply:
[Updated the comments ](https://github.com/openebs/openebs-e2e/pull/66/commits/fdcab88233301473e1f0ba4206b84d5193904e28#diff-a1f3235d393e75d18b2ad6bb955241c24fb3e8b134317f76ce9dc3bc1446f9bcR78)